### ISSUE TITLE
train_test_split  with fixed random seed gives random result

### ISSUE THREAD

**Issue by brethoehn** [RasaHQ/rasa#9238]
The method [train_test_split| https://github.com/RasaHQ/rasa/blob/3f3c8830a99278578ea687bdfc3645d7c368e86b/rasa/shared/nlu/training_data/training_data.py#L517] has an optional random_seed parameter, which is used to seed the random shuffle applied to the training examples (in the _split_cases method within the [split_nlu_examples| https://github.com/RasaHQ/rasa/blob/3f3c8830a99278578ea687bdfc3645d7c368e86b/rasa/shared/nlu/training_data/training_data.py#L567] method) -- so if we pass in the same random seed multiple times, we expect to get the same splitting of the data.

However, instead we're observing different splits, apparently because split_nlu_examples converts self.training_examples into a [Set| https://docs.python.org/3/tutorial/datastructures.html#sets] object:  _training_examples = set(self.training_examples)_ 

and this randomly reorders the training_examples (Set objects do not preserve order). The Set object is useful for automatically removing duplicates, and for performing fast Set subtraction:  _training_examples = training_examples - set(examples)_ 

I have tried adjusting the code to avoid the conversion to a Set object, and replaced the Set subtraction with list comprehension:  _training_examples = [t for t in training_examples if t not in examples]_  This has the desired result (we get the same data split when running multiple times with the same random seed), but the list comprehension can take significantly longer. Is there a better solution?

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @ArjaanBuijk will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by m-vdb** [comment]
Closing as stale
